### ISSUE TITLE
[Pallas] Add lowering support for tl.join

### ISSUE THREAD

**Issue by ggengnv** [jax-ml/jax#25321]
Pallas GPU doesn't seem to support `tl.join`. This is useful for writing kernels that involve unpacking int4 values from int8's. E.g., see XLA logic for emitting int4 unpack: [link](https://github.com/openxla/xla/blob/main/xla/service/gpu/fusions/triton/triton_fusion_emitter_legacy_matmul.cc#L513)

Perhaps `jnp.stack` could be used as the frontend, with the restriction that the number of arrays is 2 or power of 2?

**Comment by ggengnv** [comment]
@gflegar yes this is still relevant, since it now looks like join is still needed for the int4 path.

**Comment by gflegar** [comment]
Why is that? I see that https://github.com/triton-lang/triton/pull/5475 has been merged, so I imagine that the work on expanding it to i4 and exposing it in TTIR can start? (And if it's just about implementing _something_, we could likely do without `tl.join` by loading every value twice and using something like `tl.where(...,lower,upper)` to get the right matrix. It's probably not very efficient, but it's not like join+reshape was that efficient either...)

**Comment by gflegar** [comment]
Or is https://github.com/triton-lang/triton/pull/5924 I just noticed also relevant context here? Are we changing the plans about having a dedicated op for i4->i8 conversion, and still considering the `reshape(join(lower, upper))` implementation?

**Comment by ggengnv** [comment]
yes the PR you linked is relevant. basically we now still have to go thru the reshape(join) implementation. With the landing of https://github.com/triton-lang/triton/pull/5924, that path willl be close to optimal.

**Comment by superbobry** [comment]
@ggengnv I will close the issue for now. Please `jnp.stack` a go with nightly jaxlib and re-open if necessary.
